### PR TITLE
feat: generate types for `joins`

### PR DESCRIPTION
**Pull request by r1tsuu**
### What?
Generates types for `joins` property.
Example from our `joins` test, keys are type-safe:
image

Output in `payload-types.ts`:
```ts
 collectionsJoins: {
    categories: {
      relatedPosts: 'posts';
      hasManyPosts: 'posts';
      hasManyPostsLocalized: 'posts';
      'group.relatedPosts': 'posts';
      'group.camelCasePosts': 'posts';
      filtered: 'posts';
      singulars: 'singular';
    };
  };
```
Additionally, we include type information about on which collection the join is, it will help when we have types generation for `where` and `sort`.

### Why?
It provides a better DX as you don't need to memoize your keys.

### How?
Modifies `configToJSONSchema` to generate the json schema for `collectionsJoins`, uses that type within `JoinQuery`

### CODE REVIEW

File: packages/payload/src/utilities/configToJSONSchema.ts [hunk 1]
```diff
@@ -93,6 +93,41 @@ function generateEntitySelectSchemas(
   }
 }
 
+function generateCollectionJoinsSchemas(collections: SanitizedCollectionConfig[]): JSONSchema4 {
+  const properties = [...collections].reduce<Record<string, JSONSchema4>>(
```
Comment (DanRibbens) [inline packages/payload/src/utilities/configToJSONSchema.ts:97]: I don't think we would want to have all collections added to `collectionJoins` do we? I would expect this to be filtered down to only the few that have joins defined.

File: test/joins/payload-types.ts [hunk 2]
```diff
@@ -26,7 +26,39 @@ export interface Config {
     'payload-preferences': PayloadPreference;
     'payload-migrations': PayloadMigration;
   };
-  collectionsSelect?: {
+  collectionsJoins: {
+    posts: {};
```
Comment (DanRibbens) [inline test/joins/payload-types.ts:30]: It'd be nicer to not have the keys with empty objects since you can't use it, right?

**Review by DanRibbens — CHANGES_REQUESTED** [review]
Great to have the types, just wondering if we can clean it up.

File: packages/payload/src/utilities/configToJSONSchema.ts (hunk 1 above)
Comment (r1tsuu) [reply packages/payload/src/utilities/configToJSONSchema.ts:97]: Updated and now it's even better, if the collection doesn't have any joins the result type for `joins` is `never`, therefore you can't use it.

image

**Review by DanRibbens — APPROVED** [review]

**Pull request merged**